src/vidpf.rs

```diff
@@ -0,0 +1,676 @@
+                &pk.cw[level],
+                parameter,
+            );
+            pi = self.proof_next(&pi, alpha, level, &s_i, t_i, &pk.cs[level]);
+        }
+
+        OutputShare { y, pi }
+    }
+
+    /// [Vidpf::verify] checks whether the proofs are equal and that the shares add up to beta.
+    #[must_use]
+    pub fn verify(&self, a: &OutputShare<V>, b: &OutputShare<V>, beta: &V) -> bool {
+        a.y.borrow() + b.y.borrow() == *beta && a.pi.0 == b.pi.0
+    }
+
+    fn prg(&self, seed: &Seed) -> (Seed, ControlBit, Seed, ControlBit) {
+        let dst = b"100";
+        let mut rng = XofFixedKeyAes128::seed_stream(&XofSeed(*seed), dst, &self.nonce);
+
+        let mut sl = Seed::default();
+        let mut sr = Seed::default();
+        rng.fill_bytes(&mut sl);
+        rng.fill_bytes(&mut sr);
+        // Use LSB of seed for ControlBits -- This ensures [prg] costs two AES calls only.
+        let tl = ControlBit::from(sl[0] & 0x1);
+        let tr = ControlBit::from(sr[0] & 0x1);
+
+        (sl, tl, sr, tr)
+    }
+
+    fn convert(&self, seed: &Seed, parameter: &V::ValueParameter) -> (Seed, V) {
+        let dst = b"101";
+        let mut rng = XofFixedKeyAes128::seed_stream(&XofSeed(*seed), dst, &self.nonce);
+
+        let mut out_seed = Seed::default();
+        rng.fill_bytes(&mut out_seed);
+        let value = V::generate(&mut rng, parameter);
+
+        (out_seed, value)
+    }
+
+    fn hash_one(alpha: &VidpfInput, level: usize, seed: &Seed) -> Proof {
+        let dst = b"vidpf cs proof";
+        let mut nonce = Vec::new();
+        nonce.extend_from_slice(&alpha.prefix(level).to_bytes());
+        nonce.extend(u32::to_le_bytes(level.try_into().unwrap()));
+        let mut rng = XofTurboShake128::seed_stream(&XofSeed(*seed), dst, &nonce);
+
+        let mut proof = Proof::default();
+        rng.fill_bytes(&mut proof.0);
+
+        proof
+    }
+
+    fn hash_two(proof: &Proof) -> Proof {
+        let dst = b"vidpf proof adjustment";
+        let nonce = &proof.0;
+        let seed = Seed::default();
+        let mut rng = XofTurboShake128::seed_stream(&XofSeed(seed), dst, nonce);
+
+        let mut out_proof = Proof::default();
+        rng.fill_bytes(&mut out_proof.0);
+
+        out_proof
+    }
+}
+
+/// VIDPF-related errors.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum VidpfError {
+    /// Invalid parameter, indicates an invalid input.
+    #[error("invalid parameter: {0}")]
+    InvalidParameter(String),
+
+    /// Failure when calling getrandom().
+    #[error("getrandom: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// `ServerId` used to identify the two aggregation servers.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum ServerId {
+    /// S0 is the first server.
+    S0 = 0,
+    /// S1 is the second server.
+    S1 = 1,
+}
+
+/// OutputShare is a share of the output value produced during the evaluation tree.
+#[derive(Debug)]
+pub struct OutputShare<V>
```

Comment:
I think this should be called a `ValueShare` in order to avoid conflating this with the concept of a "VDAF output share".